crd/crd_test.go

```diff
@@ -104,37 +102,45 @@ func functionTests(crdClient *rest.RESTClient) {
 	panicIf(err)
 
 	// start a watch
+
+	// Kubernetes watch returns a sequence changes of CRD resources.
+	// In some cases, the older changes will be returned and breaks the test.
+	// To prevent this, we wait for a couple of seconds to ensure we get all
+	// changes from the server and then verify the content of the object.
+
 	wi, err := fi.Watch(metav1.ListOptions{})
 	panicIf(err)
 
 	start := time.Now()
 	function.Metadata.ResourceVersion = ""
 	f, err = fi.Create(function)
 	panicIf(err)
-	defer fi.Delete(f.Metadata.Name, nil)
+	defer func() {
+		fi.Delete(f.Metadata.Name, nil)
+		wi.Stop()
+	}()
 
 	// assert that we get a watch event for the new function
-	recvd := false
-	select {
-	case <-time.NewTimer(1 * time.Second).C:
-		if !recvd {
-			log.Panicf("Didn't get watch event")
-		}
-	case ev := <-wi.ResultChan():
-		wf, ok := ev.Object.(*Function)
-		if !ok {
-			log.Panicf("Can't cast to Function")
-		}
-		if wf.Spec.Environment.Name != function.Spec.Environment.Name {
-			log.Panicf("Bad object from watch: %#v", wf)
+	var watchFunc *Function
+	for {
+		select {
+		case <-time.NewTimer(3 * time.Second).C:
+			if watchFunc.Spec.Environment.Name != function.Spec.Environment.Name {
```

Comment:
1. check for watchFunc == nil?
 2. I'm confused about the change, overall.  Are you trying to ignore stale changes?  If so why error out on them?  Also what do we gain from changing the timer from 1 sec to 3 sec?